### PR TITLE
merge my branch prune

### PR DESCRIPTION
**Pull request by Alejandro-Casanova**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces a new script and functionality for model pruning in YOLOv8, along with various configuration adjustments and enhancements focused on improving the pruning process.

### 📊 Key Changes
- **.gitignore Update:** Added `nohup.out` to the ignore list.
- **Pruning Results Storage:** Added files for storing pruning performance metrics (`pruning_perf_change_0.png`, `pruning_perf_change_0.txt`, etc.).
- **New Pruning Script:** Introduced `yolov8_pruning.py` for handling model pruning tasks.
- **Configuration File:** Added `ultralytics/cfg/my_test_cfg.yaml` for specifying training settings and parameters.
- **Loss Function Adjustments:** Minor changes in `ultralytics/utils/loss.py` for improved functionality.
- **Test Script:** Included `test_run.sh` for running pruning tests.

### 🎯 Purpose & Impact
- **Streamlined Pruning:** Allows for iterative and fine-tuning pruning of YOLOv8 models to improve performance while reducing model size.
- **Better Resource Management:** Pruning helps in reducing computational requirements, making models more efficient for deployment.
- **Configurability:** New configuration files provide granular control over training and pruning parameters, enhancing flexibility for users.
- **Ease of Use:** Added scripts and utilities simplify the process of running and validating pruning operations.

Overall, these changes aim to make YOLOv8 more efficient, resource-optimized, and user-friendly, especially in environments with constrained computational resources. 🚀

### CODE REVIEW

**Comment by Alejandro-Casanova** [comment]
sorry, this was a mistake hahahaha

**Comment by Ramzes30765** [comment]
@Alejandro-Casanova does your pruning working?

**Comment by Alejandro-Casanova** [comment]
> @Alejandro-Casanova does your pruning working?

not really, still working on it

**Comment by glenn-jocher** [comment]
Hi @Alejandro-Casanova,

Thank you for your update! If you encounter any specific issues or need assistance with the pruning process, please feel free to share more details. Providing a minimum reproducible example can be very helpful for us to understand and address the problem effectively. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

Additionally, ensure that you are using the latest version of the Ultralytics packages, as updates often include important fixes and improvements.

Looking forward to your progress! 🚀

**Comment by Ramzes30765** [comment]
> > @Alejandro-Casanova does your pruning working?
> 
> not really, still working on it

If it is possible tag me please if your version will work

**Comment by Alejandro-Casanova** [comment]
> > > @Alejandro-Casanova does your pruning working?
> > 
> > 
> > not really, still working on it
> 
> If it is possible tag me please if your version will work

sure! I'll tell you if I get it working!

**Comment by Alejandro-Casanova** [comment]
> Hi @Alejandro-Casanova,
> 
> Thank you for your update! If you encounter any specific issues or need assistance with the pruning process, please feel free to share more details. Providing a minimum reproducible example can be very helpful for us to understand and address the problem effectively. You can find guidelines on how to create one here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).
> 
> Additionally, ensure that you are using the latest version of the Ultralytics packages, as updates often include important fixes and improvements.
> 
> Looking forward to your progress! 🚀

Would you be able to tell me how the training process was for the pre-trained yolov8 models? How the hyper-parameters were set (epochs, lr, data augmentation...). Is this documented somewhere? I want to fine-tune my model after pruning, and I'd like to mimic the pre-training procedure, as close as possible, but with reduced epochs and learning rate. Thank you in advance for your help!

**Comment by glenn-jocher** [comment]
Hi there,

Thank you for your question! The training process for the pre-trained YOLOv8 models involves a comprehensive setup of hyperparameters and data augmentation techniques to ensure optimal performance. While the exact details of the training process for the pre-trained models are proprietary, I can provide you with a general guideline to help you fine-tune your model effectively.

### General Training Setup for YOLOv8:

1. **Epochs**: Typically, models are trained for 300-500 epochs, but for fine-tuning, you might reduce this to 50-100 epochs depending on your dataset size and complexity.
2. **Learning Rate**: Start with a lower learning rate for fine-tuning. If the original training used a learning rate of `0.01`, you might want to start with `0.001` or `0.0001`.
3. **Data Augmentation**: Common augmentations include random scaling, flipping, rotation, and color jittering. These help in making the model robust to various transformations.
4. **Batch Size**: A batch size of 16 or 32 is commonly used, but this can be adjusted based on your GPU memory.
5. **Optimizer**: Adam or SGD with momentum are popular choices. Fine-tuning often benefits from using the same optimizer as the initial training.

### Example Code for Fine-Tuning:

Here's a sample code snippet to help you get started with fine-tuning your pruned model:

```python
from ultralytics import YOLO

# Load your pruned model
model = YOLO('path/to/pruned_model.pt')

# Fine-tune the model
results = model.train(
    data='path/to/your_dataset.yaml',
    epochs=50,  # Reduced epochs for fine-tuning
    lr0=0.001,  # Lower learning rate
    imgsz=640,  # Image size
    batch=16,  # Batch size
    augment=True  # Enable data augmentation
)
```

### Documentation and Resources:

For more detailed information on training and fine-tuning, you can refer to the [Ultralytics YOLOv8 documentation](https://docs.ultralytics.com/models/yolov8). This resource provides comprehensive guidelines on various training parameters and best practices.

Feel free to reach out if you have any more questions or need further assistance. Happy fine-tuning! 🚀

**Comment by Alejandro-Casanova** [comment]
> Hi there,
> 
> Thank you for your question! The training process for the pre-trained YOLOv8 models involves a comprehensive setup of hyperparameters and data augmentation techniques to ensure optimal performance. While the exact details of the training process for the pre-trained models are proprietary, I can provide you with a general guideline to help you fine-tune your model effectively.
> ### General Training Setup for YOLOv8:
> 
>     1. **Epochs**: Typically, models are trained for 300-500 epochs, but for fine-tuning, you might reduce this to 50-100 epochs depending on your dataset size and complexity.
> 
>     2. **Learning Rate**: Start with a lower learning rate for fine-tuning. If the original training used a learning rate of `0.01`, you might want to start with `0.001` or `0.0001`.
> 
>     3. **Data Augmentation**: Common augmentations include random scaling, flipping, rotation, and color jittering. These help in making the model robust to various transformations.
> 
>     4. **Batch Size**: A batch size of 16 or 32 is commonly used, but this can be adjusted based on your GPU memory.
> 
>     5. **Optimizer**: Adam or SGD with momentum are popular choices. Fine-tuning often benefits from using the same optimizer as the initial training.
> 
> 
> ### Example Code for Fine-Tuning:
> 
> Here's a sample code snippet to help you get started with fine-tuning your pruned model:
> 
> ```python
> from ultralytics import YOLO
> 
> # Load your pruned model
> model = YOLO('path/to/pruned_model.pt')
> 
> # Fine-tune the model
> results = model.train(
>     data='path/to/your_dataset.yaml',
>     epochs=50,  # Reduced epochs for fine-tuning
>     lr0=0.001,  # Lower learning rate
>     imgsz=640,  # Image size
>     batch=16,  # Batch size
>     augment=True  # Enable data augmentation
> )
> ```
> 
> ### Documentation and Resources:
> 
> For more detailed information on training and fine-tuning, you can refer to the [Ultralytics YOLOv8 documentation](https://docs.ultralytics.com/models/yolov8). This resource provides comprehensive guidelines on various training parameters and best practices.
> 
> Feel free to reach out if you have any more questions or need further assistance. Happy fine-tuning! 🚀

thank you!

**Comment by glenn-jocher** [comment]
You're welcome! 😊 If you have any more questions or run into any issues, feel free to ask. We're here to help you get the most out of your YOLOv8 model. Happy fine-tuning and best of luck with your project! 🚀
